README.md
# st-react-autosuggest

## Description
This is a heavily abstracted autosuggest built on 'st-react-typeahead-component'. All properties passed to this component will be passed down to the underlying autosuggest except for 'suggestions' as it is going to be provided by out backing AutosuggestStore.

## Updating st-react-typeahead-component
This component depends on st-react-typeahead-component which is just [a forked repo](https://github.com/fumblesandfriends/react-typeahead-component) that must be manually published to npm.

To get artifactory access see the `Publishing NPM Module`

- Checkout the repo
- Make changes and change versions
- Commit
- Apply diff below
- run `npm publish`

### Patch file

```text

diff --git a/package.json b/package.json
index e5a5756..2dfe69c 100644
--- a/package.json
+++ b/package.json
@@ -1,5 +1,5 @@
 {
-  "name": "@tappleby/react-typeahead-component",
+  "name": "st-react-typeahead-component",
   "description": "Typeahead, written using the React.js library.",
   "author": "Ezequiel Rodriguez <ezequiel@yahoo.com>",
   "version": "0.10.0-alpha2",
@@ -10,6 +10,9 @@
     "type": "git",
     "url": "https://github.com/ezequiel/react-typeahead-component.git"
   },
+  "publishConfig": {
+    "registry": "https://artifactory.smartertravel.net/artifactory/api/npm/npm-local"
+  },
   "keywords": [
     "react",
     "reactjs",

```

##Usage
This autosuggest requires an implementation of 'src/lib/store/autosuggestStore' passed as a property.

### IE9 and FF Fixes
IE9 has an issue where the disabled input still gets the click / focus event even if it is underneath the
active input. FF has an issue where the active input is not positioned correctly. Here is some css that
fixes the issues:

```css
input.react-typeahead-input.react-typeahead-hint {
  opacity: 1;
  z-index: 0;
  left: -9999999999px;
}
input.react-typeahead-input.react-typeahead-usertext {
  top: 0;
  right: 0;
  z-index: 1;
  position: absolute !important;
  background-color: #fff !important;
}
```

#### OptionTemplate
For what is available to the option template see the [readme](https://github.com/fumblesandfriends/react-typeahead-component/blob/master/README.md#reactelement-optiontemplate-required)

```js

import React from "react";
export default class OptionsTemplate extends React.Component {
  render() {
      var classes = "option-value",
      optionData = this.props.data;
      if (this.props.isSelected) {
        classes += " selected-option";
      }
      return (
          <div>
              <div className={classes}>
                  {optionData}
              </div>
          </div>
      );
  }
}
OptionsTemplate.propTypes = {
  data: React.PropTypes.string,
  isSelected: React.PropTypes.bool
};

```

#### AutosuggestStore

```js

import AutosuggestStore from "./autosuggestStore";
import OptionTemplate from "../ui/static-option-template";

class AutosuggestStaticStore extends AutosuggestStore {

  constructor(locations = ["Boston", "New York", "Miami"]) {
    super();
    this.locations = locations;
  }

  getSuggestion(suggestion, callback) {
    const regex = new RegExp("^" + suggestion, "i");
    const suggestions = this.locations.filter(suburb => regex.test(suburb));
    setTimeout(() => callback(null, suggestions), 300); // Emulate API call
  }

  getSuggestionTemplate() {
    return OptionTemplate;
  }

  getDisplayValue(suggestion) {
    return suggestion;
  }
}

export default AutosuggestStaticStore;

```

## AutosuggestUI

```js

import {AutosuggestUI} from "st-react-autosuggest"
import AutosuggestStaticStore from ...
<AutosuggestUI autosuggestStore={new AutosuggestStaticStore()} />

```

###Options

```js

/**
 * @attribute autosuggestStore
 * @required
 * @type {AutosuggestStore}
 */
autosuggestStore: React.PropTypes.any.isRequired,

/**
 * @attribute initialSuggestion
 * @type Suggestion
 */
initialSuggestion: React.PropTypes.any,

/**
 * @attribute focusPlaceholderText
 * @type String
 * @default ""
 */
focusPlaceholderText: React.PropTypes.string,

/**
 * @attribute onSelected
 * @type Function
 *   @param {Suggestion}
 *
 */
onSelected: React.PropTypes.func

```


## Developing

Pull down all the devDependencies

```
npm install
```

To see the list of available grunt tasks

```
grunt
```

The watch task:

```bash
grunt watch
```
As you make changes to the code your files will be automatically linted and the unit tests will be run.

### Generate docs
To generate yui docs at http://sites.smartertravel.net/projects/st-react-datepicker/

```bash
npm run docs
```

## Publishing NPM Module
While the Grunt task will take care of publishing the NPM module, you would still need to setup the credentials on your
machine to actually publish it. Begin, by adding the authentication token to your `.npmrc` file (this should be the one
in your home directory). You can do that by running the command similar to the following:

```bash
curl -u{user}:{password} "https://artifactory.smartertravel.net/artifactory/api/npm/auth" >> ~/.npmrc
```

Note that your password needs to be the encrypted version. To get that encrypted version of your password:
* Log into Artifactory thru the UI and you to your user's profile.
* Type in your password again and click on Unlock.
* The field below with your encrypted password will now be shown which you can copy and paste into the command above.

Once you have your authentication setup, you can now push to Artifactory.

---


CHANGELOG.md
## 2.0.4 (2016-01-25)

- Fixing dependency issue

## 2.0.3 (2016-01-25)

- Fixed issue where focus does not clear the input directly after an option was selected

## 2.0.2 (2016-01-22)

- Updated st-react-typeahead-component: IE Fixes where the focus event breaks the onOptionClick function
- Added flag to prevent clearing of input on focus when a new option was just selected

## 2.0.0 (2016-01-21)

- Changed underlying autosuggest implementation to st-react-typeahead-component
- Added keyboard support
- Added input clear functionality on focus and revert on blur
- Closing the selector will select the last focused option

## 1.1.0 (2016-01-19)



## 1.0.0 (2015-12-01)

- Update to react 0.14.0


test/autosuggest-test.js
import React from "react";
import TestUtils from "react-addons-test-utils";
import ReactDOM from "react-dom";
import chai, {expect} from "chai";
import sinon from "sinon";
import sinonChai from "sinon-chai";
import {AutosuggestUI} from "../src/index";
import AutosuggestStaticStore from "../src/lib/store/autosuggestStaticStore";
import Autosuggest from "st-react-typeahead-component";
chai.use(sinonChai);

describe("Autosuggest UI", function () {
  it("Should fail to render", function () {
    const shallowRenderer = TestUtils.createRenderer();
    try {
      shallowRenderer.render(React.createElement(AutosuggestUI));
      return false;
    } catch (e) {
      return true;
    }
  });

  it("Should render correct underlying autosuggest lib", function () {
    const shallowRenderer = TestUtils.createRenderer();
    shallowRenderer.render(React.createElement(AutosuggestUI, {"autosuggestStore": new AutosuggestStaticStore()}));
    var output = shallowRenderer.getRenderOutput();
    expect(output.props.children.type).to.equal(Autosuggest);
  });

  it("Should render with initialSuggestion", function () {
    var div = document.createElement('div');
    var typeaheadInstance = ReactDOM.render(
            <AutosuggestUI
                initialSuggestion='Miami'
                autosuggestStore={new AutosuggestStaticStore()}
            />,
            div
        );

    expect(typeaheadInstance.state.input).to.equal("Miami");
    expect(typeaheadInstance.state.options).to.deep.equal([]);
  });

  it("Focus should clear current value and blur should restore it", function () {
    var div = document.createElement('div');
    var typeaheadInstance = ReactDOM.render(
            <AutosuggestUI
                initialSuggestion='Miami'
                autosuggestStore={new AutosuggestStaticStore()}
            />,
            div
        );

    // The hint should be visible at this point.
    expect(typeaheadInstance.state.input).to.equal("Miami");
    expect(typeaheadInstance.state.options).to.deep.equal([]);

    var input = TestUtils.findRenderedDOMComponentWithClass(typeaheadInstance, "react-typeahead-usertext");
    TestUtils.Simulate.focus(input);

    expect(typeaheadInstance.state.input).to.equal("");
    expect(typeaheadInstance.state.options).to.deep.equal([]);

    TestUtils.Simulate.blur(input);

    expect(typeaheadInstance.state.input).to.equal("Miami");
    expect(typeaheadInstance.state.options).to.deep.equal([]);

  });


});


test/autosuggestStaticStore-test.js
import {expect} from "chai";
import AutosuggestStaticStore from "../src/lib/store/autosuggestStaticStore";
import StaticOptionTemplate from "../src/lib/ui/static-option-template";

describe("AutosuggestStaticStore", function () {
  it("callback should be called with array containing boston", function () {
    var cb = function (arg1, arg2) {
      expect(arg1).to.equal(null);
      expect(arg2).to.equal(["Boston"]);
      done();
    };

    var autosuggestStore = new AutosuggestStaticStore();
    autosuggestStore.getSuggestion("bos", cb);
  });

  it("callback should be called with an empty array", function () {
    var cb = function (arg1, arg2) {
      expect(arg1).to.equal(null);
      expect(arg2).to.equal([]);
      done();
    };

    var autosuggestStore = new AutosuggestStaticStore(["Miami"]);
    autosuggestStore.getSuggestion("bos", cb);
  });

  it("getSuggestionTemplate should return the static option template", function () {
    var autosuggestStore = new AutosuggestStaticStore(["Miami"]);
    expect(autosuggestStore.getSuggestionTemplate()).to.equal(StaticOptionTemplate);
  });

  it("getDisplayValue should return the suggestion", function () {
    var autosuggestStore = new AutosuggestStaticStore(["Miami"]);
    expect(autosuggestStore.getDisplayValue("Miami")).to.equal("Miami");
  });
});


grunt/deploy.js
var _ = require("lodash");
/**
 * This deploys the npm modules to artifactory see README.md for prereqs
 * See documentation below for more details
 */
module.exports = function(grunt, configs) {

  /**
   * Based on the flags that were given, return a two-item array where the first item is the name of the bump task to
   * run and the second item being whether or not the task should run.
   *
   * @param {String[]} flags
   * @return Array
   */
  function getBumpTaskConfigs(flags) {
    if (_.contains(flags, "disableBump")) {
      return ["bump", false];
    }

    var bumpConfig = _.find(flags, function(flag) { return _.startsWith(flag, "enableBump("); });
    if (!bumpConfig) {
      return ["bump:patch", true];
    }

    var versionType = bumpConfig.slice(11, -1);
    return ["bump:" + versionType, true];
  }

  /**
   * Task to tag and deploy the current state to git and npm.
   *
   * This task will take care of running other tasks in the right order, specified below:
   *     - clean:npm - Enabled by default, can be disabled with "disableClean".
   *     - test - Enabled by default, can be disabled with "disableTests".
   *     - run:babel - Will always run. This transpiles and code before publishing.
   *     - bump:{versionType} - Enabled by default with the versionType being "patch". To specify a different type,
   *         use "enableBump(versionType)". To disable, use "disableBump".
   *     - package:npm - Will always run.
   *     - publish:npm - Enabled by default, can be disabled with "disablePublish".
   *
   * Below are some examples of how this task would run:
   *     - internal-deploy
   *       Will run the tasks "clean:npm", "test" "bump:patch", "copy:npm", "package:npm", and "publish:npm".
   *     - internal-deploy:disableClean:enableJsHint:enableBump(minor):disablePublish
   *       Will run the tasks "test", "bump:minor", "copy:npm", "package:npm"
   *     - internal-deploy:disableBump
   *       Will run the tasks "clean:npm", "copy:npm", "package:npm", "publish:npm".
   */
  grunt.registerTask(
    "internal-deploy",
    "Update the version number, create a tag for it, and push changes to git and npm.",
    function() {
      var flags = _.toArray(arguments),
        tasks = [
          ["clean:npm", !_.contains(flags, "disableClean")],
          ["test", !_.contains(flags, "disableTests")],
          ["run:babel", true],
          getBumpTaskConfigs(flags),
          ["package:npm", true],
          ["publish:npm", !_.contains(flags, "disablePublish")]
        ];

      _.each(tasks, function(taskConfigs) {
        var taskName = taskConfigs[0],
          shouldRun = taskConfigs[1];

        if (shouldRun) {
          grunt.task.run(taskName);
        }
      });
    }
  );

  grunt.registerTask(
    "deploy:dev",
    "Clean, verify JavaScript, update the pre-patch version number, create tag for it, and push to git & npm.",
    "internal-deploy:enableJsHint:enableBump(prepatch)"
  );
  grunt.registerTask(
    "deploy:major",
    "Clean, verify JavaScript, update the major version number, create tag for it, and push to git & npm.",
    "internal-deploy:enableBump(major)"
  );
  grunt.registerTask(
    "deploy:minor",
    "Clean, verify JavaScript, update the minor version number, create tag for it, and push to git & npm.",
    "internal-deploy:enableBump(minor)"
  );
  grunt.registerTask(
    "deploy:patch",
    "Clean, verify JavaScript, update the patch version number, create tag for it, and push to git & npm.",
    "internal-deploy"
  );

  // This task doesn't have any configurations
  return null;
};


src/lib/store/autosuggestStore.js
/**
 * @module st-react-autosuggest
 */
/**
 * The base class that any autosuggest store needs to override. This is an 'interface'.
 * The autosuggest store should take a locale in the constructor to localize suggestions.
 * @class AutosuggestStore
 */
class AutosuggestStore {
  constructor() {}
  /**
   * This function takes a suggestion and passes an arry of possible matches to
   * the provided callback
   * @method getSuggestion
   * @param {String} suggestion the suggestion to look up
   * @param {Function} callback function that is called when results are recieved.
   *  @param {Error} callback.error an error if it occured
   *  @param {Array} callback.suggestions an array of suggestions
   *    @param {Suggestion} callback.suggestions.suggestion  A suggestions is String, e.g.: 'Mentone' or Object, e.g.: { suburb: 'Mentone', postcode: '3194' }
   */
  getSuggestion(suggestion, callback) {
    callback(new Error("Method not implemented"), null);
  }

  /**
   * @method getSuggestionValue
   * @param {Suggestion} suggestion
   * @return {React.Component} see https://github.com/ezequiel/react-typeahead-component#reactelement-optiontemplate-required
   */
  getSuggestionTemplate() {
    throw "Method not implemented";
  }

  /**
   * @method getDisplayValue
   * @param {Suggestion} suggestion
   * @return {String} the display value for the suggestion
   */
  getDisplayValue(Suggestion) {
    throw "Method not implemented";
  }
}

export default AutosuggestStore;


src/lib/store/autosuggestStaticStore.js
/**
 * @module st-react-autosuggest
 */
import AutosuggestStore from "./autosuggestStore";
import OptionTemplate from "../ui/static-option-template";

/**
 * This is a sample implementation of an AutosuggestStore. This should never be used in
 * production. This is currently only used for unit tests.
 * @class AutosuggestStaticStore
 * @extends AutosuggestStore
 * @constructor
 * @param {Array} locations the list of suggestable locations.
 *  @param {String} locations.location A string name "Boston"
 * @example
 *   var store = new AutosuggestStaticStore(["Boston", "Miami"]);
 *
 */
class AutosuggestStaticStore extends AutosuggestStore {

  constructor(locations = ["Boston", "New York", "Miami"]) {
    super();
    this.locations = locations;
  }

  /**
   * This function takes a suggestion and passes an arry of possible matches to
   * the provided callback
   * @method getSuggestion
   * @param {String} suggestion the suggestion to look up
   * @param {Function} callback function that is called when results are recieved.
   *  @param {Error} callback.error an error if it occured
   *  @param {Array} callback.suggestions an array of suggestions
   *    @param {Suggestion} callback.suggestions.suggestion  A suggestions is String, e.g.: 'Mentone' or Object, e.g.: { suburb: 'Mentone', postcode: '3194' }
   */
  getSuggestion(suggestion, callback) {
    const regex = new RegExp("^" + suggestion, "i");
    const suggestions = this.locations.filter(suburb => regex.test(suburb));
    setTimeout(() => callback(null, suggestions), 300); // Emulate API call
  }

  getSuggestionTemplate() {
    return OptionTemplate;
  }

  getDisplayValue(suggestion) {
    return suggestion;
  }
}

export default AutosuggestStaticStore;
